library/core/src/ptr/mod.rs

```diff
@@ -59,11 +59,39 @@
 //!
 //! ## Allocated object
 //!
-//! For several operations, such as [`offset`] or field projections (`expr.field`), the notion of an
-//! "allocated object" becomes relevant. An allocated object is a contiguous region of memory.
-//! Common examples of allocated objects include stack-allocated variables (each variable is a
-//! separate allocated object), heap allocations (each allocation created by the global allocator is
-//! a separate allocated object), and `static` variables.
+//! An *allocated object* is a subset of program memory which is addressable
+//! from Rust, and within which pointer arithmetic is possible. Examples of
+//! allocated objects include heap allocations, stack-allocated variables,
+//! statics, and consts. The safety preconditions of some Rust operations -
+//! such as `offset` and field projections (`expr.field`) - are defined in
+//! terms of the allocated objects on which they operate.
+//!
+//! An allocated object has a base address, a size, and a set of memory
+//! addresses. It is possible for an allocated object to have zero size, but
+//! such an allocated object will still have a base address. The base address
+//! of an allocated object is not necessarily unique. While it is currently the
+//! case that an allocated object always has a set of memory addresses which is
+//! fully contiguous (i.e., has no "holes"), there is no guarantee that this
+//! will not change in the future.
+//!
+//! For any allocated object with `base` address, `size`, and a set of
+//! `addresses`, the following are guaranteed:
+//! - For all addresses `a` in `addresses`, `a` is in the range `base .. (base +
+//!   size)` (note that this requires `a < base + size`, not `a <= base + size`)
+//! - `base` is not equal to [`null()`] (i.e., the address with the numerical
+//!   value 0)
+//! - `base + size <= usize::MAX`
+//! - `size <= isize::MAX`
+//!
+//! As a consequence of these guarantees, given any address, `a`, within the set
+//! of addresses of an allocated object:
+//! - It is guaranteed that `a - base` does not overflow `isize`
+//! - It is guaranteed that `a - base` is non-negative
```

Comment:
Doesn't the latter imply the former? Both a and base are positive, so the only way they could overflow is by becoming really negative.

Reply:
I *think* I agree that the latter implies the former, but I'd have to think more about it to convince myself. However, a lot of pointer APIs explicitly mention the "does not overflow `isize`" requirement, so IMO it's worth stating explicitly so that callers who are writing safety comments can just reference this directly rather than having to have their safety comments contain more complex logic. I grant that it's not *that* complex, but from the perspective of someone who writes a lot of safety comments, even relatively simple logic is way more annoying (and error prone) than just being able to directly cite a documented guarantee.

Reply:
> and also for there to be a section of things that aren't implications of the rules and thus cannot generally be relied upon.

Interesting thought. Any come to mind that we should disclaim here?

Reply:
Yeah, agreed that that probably makes more sense for the next section. But I like the pattern!

Reply:
> I think that it's reasonable for the "As a consequence of these guarantees" section to contain things that are "just" implications

Okay that's fair.